agent/resources_monitor.go

```diff
@@ -0,0 +1,135 @@
+package agent
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/agent/proto"
+	"github.com/coder/coder/v2/cli/clistat"
+	"github.com/coder/quartz"
+)
+
+func (a *agent) pushResourcesMonitoring(ctx context.Context, aAPI proto.DRPCAgentClient24) error {
+	logger := a.logger.Named("resources_monitor")
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	clk := quartz.NewReal()
+
+	return PushResourcesMonitoringWithConfig(ctx, logger, clk, aAPI.GetResourcesMonitoringConfiguration, aAPI.PushResourcesMonitoringUsage)
+}
+
+type (
+	ResourcesMonitorConfigurationFetcher func(ctx context.Context, params *proto.GetResourcesMonitoringConfigurationRequest) (*proto.GetResourcesMonitoringConfigurationResponse, error)
+	ResourcesMonitorDatapointsPusher     func(ctx context.Context, params *proto.PushResourcesMonitoringUsageRequest) (*proto.PushResourcesMonitoringUsageResponse, error)
+)
+
+func PushResourcesMonitoringWithConfig(ctx context.Context,
+	logger slog.Logger,
+	clk quartz.Clock,
+	configFetcher ResourcesMonitorConfigurationFetcher,
+	datapointsPusher ResourcesMonitorDatapointsPusher,
+) error {
+	config, err := configFetcher(ctx, &proto.GetResourcesMonitoringConfigurationRequest{})
+	if err != nil {
+		return xerrors.Errorf("failed to get resources monitoring configuration: %w", err)
+	}
+
+	if !config.Enabled {
+		logger.Info(ctx, "resources monitoring is disabled")
+		return nil
+	}
+
+	resourcesFetcher, err := clistat.New()
+	if err != nil {
+		return xerrors.Errorf("failed to create resources fetcher: %w", err)
+	}
+
+	datapointsQueue := NewResourcesMonitorQueue(int(config.Config.NumDatapoints))
+
+	clk.TickerFunc(ctx, time.Duration(config.Config.TickInterval*int32(time.Second)), func() error {
```

Comment:
self review : We prefer to log instead of returning an error in case of problem.

Using `TickerFunc` from quartz, returning an error would just shutdown the whole ticker and stop it. Returning nil stops the current tick, but the next one will still trigger.